lib/puppet/type/file.rb

```diff
@@ -904,8 +915,15 @@ def fail_if_checksum_is_wrong(path, content_checksum)
   # write the current content. Note that if there is no content property
   # simply opening the file with 'w' as done in write is enough to truncate
   # or write an empty length file.
-  def write_content(file)
-    (content = property(:content)) && content.write(file)
+  def write_contents(file)
+    # Ideally only source or content would exist. However, if source is specified without
+    # checksum_value, source will be used to generate content; content will trigger
+    # writing the file, but we should still use source to do the final write, so put it first.
+    if source = parameter(:source)
+      source.write(file)
+    elsif content = property(:content)
+      content.write(file)
+    end
```

Comment:
Since this method is called by either `content` or `checksum_value`, can the caller specify which property to use so that we don't have to guess?